nilearn/masking.py

```diff
@@ -607,32 +612,44 @@ def compute_brain_mask(target_img, threshold=.5, connected=True,
         Controls the amount of verbosity: higher numbers give
         more messages
 
+    mask_type : {'whole-brain', 'gm', 'wm'}, optional
+        Type of mask to be computed. Default = 'whole-brain'
+
     Returns
     -------
     mask : nibabel.Nifti1Image
         The whole-brain mask (3D image)
     """
     if verbose > 0:
-        print("Template mask computation")
+        msg_to_print = "Template " + mask_type + " mask computation"
+        print(msg_to_print)
 
     target_img = _utils.check_niimg(target_img)
 
-    from .datasets import load_mni152_brain_mask
-    template = load_mni152_brain_mask()
+    if mask_type == 'whole-brain':
+        template = load_mni152_brain_mask()
+    elif mask_type == 'gm':
+        template = load_mni152_gm_mask()
+    elif mask_type == 'wm':
+        template = load_mni152_wm_mask()
+    else:
+        raise ValueError("Unknown mask type {}.".format(mask_type))
+
     dtype = img_data_dtype(target_img)
     template = new_img_like(template,
                             get_data(template).astype(dtype))
 
-    from .image.resampling import resample_to_img
-    resampled_template = cache(resample_to_img, memory)(template, target_img)
+    resampled_template = cache(resampling.resample_to_img, memory)(
+        template, target_img, force_resample=True)
```

Comment:
But then we should do the thresholding at 0.2 after resampling rather than before: the continuous gm/wm has to be resampled, and the resulting image is thresholded.

Reply:
OK. I have changed the default threshold of the functions `nilearn/datasets/struct.py::load_mni152_gm_mask` and `nilearn/datasets/struct.py::load_mni152_wm_mask` from 0.2 to 0, as in `nilearn/datasets/struct.py::load_mni152_brain_mask`. Then, I changed the threshold of `nilearn/masking.py::compute_brain_mask` from 0.5 to 0.2. Do you agree with this change?

Reply:
This is not a matter of thresholds, but of operation order. If I am not mistaken, the image that is resampled here has just been binarized. It should be resampled before it is binarized.

Reply:
Thanks @bthirion, this explains why the resampling is done with continuous interpolation here. 
This means that there is currently a bug on master since `load_mni152_brain_mask()` is used and returns a binary image:

https://github.com/nilearn/nilearn/blob/a36c0442fc8eebe401b310dd9799c0f071080e90/nilearn/masking.py#L621-L627

@alpinho, I think we should load the template like this:

```python
if mask_type == 'whole-brain':
         template = load_mni152_template()
     elif mask_type == 'gm':
         template = load_mni152_gm_template()
     elif mask_type == 'wm':
         template = load_mni152_wm_template()
```

and ignore my review comments about nearest interpolation and hard thresholding...

Reply:
Shall I reverse the changes I describe in my previous comment on this thread or shall I keep them?

Reply:
I'm OK to address the bad ordering of thresholding/resampling in another PR (please open an issue though).
Best,

Reply:
I reversed these changes.

Reply:
AFAICT this has been resolved? now it seems the (floating-point) template is resampled to the `target_img`, _then_ thresholded (what @bthirion was asking for IIUC)